plot_app/configured_plots.py

```diff
@@ -141,7 +141,82 @@ def generate_plots(ulog, px4_ulog, db_data, vehicle_data, link_to_3d_page):
+
+                    color_str = flight_modes_table[flight_mode][1][1:] # color in form 'ff00aa'
+
+                    # increase brightness to match colors with template
+                    rgb = [int(color_str[2*x:2*x+2], 16) for x in range(3)]
+                    for i in range(3):
+                        rgb[i] += 40
+                        if rgb[i] > 255: rgb[i] = 255
+
+                    color_str = "".join(map(lambda x: format(x, '02x'), rgb))
+
+                    return '#'+color_str
+
+                def ulog_to_polyline(ulog):
+                    # get flight modes
+                    try:
+                        cur_dataset = ulog.get_dataset('vehicle_status')
+                        flight_mode_changes = cur_dataset.list_value_changes('nav_state')
+                        flight_mode_changes.append((ulog.last_timestamp, -1))
+                    except (KeyError, IndexError) as error:
+                        flight_mode_changes = []
+                    if flight_mode_changes is None:
+                        flight_mode_changes = []
+                    topic_instance = 0
+                    position_topic_name = 'vehicle_global_position'
+                    for elem in data:
+                        if elem.name == position_topic_name and elem.multi_id == topic_instance:
+                            cur_data = elem
+
+                    pos_lon = cur_data.data['lon']
+                    pos_lat = cur_data.data['lat']
+                    pos_alt = cur_data.data['alt']
+                    pos_t = cur_data.data['timestamp']
+
+                    if 'fix_type' in cur_data.data:
+                        indices = cur_data.data['fix_type'] > 2  # use only data with a fix
+                        pos_lon = pos_lon[indices]
+                        pos_lat = pos_lat[indices]
+                        pos_alt = pos_alt[indices]
+                        pos_t = pos_t[indices]
+
+                    # scale if it's an integer type
+                    lon_type = [f.type_str for f in cur_data.field_data if f.field_name == 'lon']
+                    if len(lon_type) > 0 and lon_type[0] == 'int32_t':
+                        pos_lon = pos_lon / 1e7  # to degrees
+                        pos_lat = pos_lat / 1e7
+                        pos_alt = pos_alt / 1e3  # to meters
+
+                    current_flight_mode = 0
+                    current_flight_mode_idx = 0
+                    if len(flight_mode_changes) > 0:
+                        current_flight_mode = flight_mode_changes[0][1]
+                    last_t = 0
+                    minimum_interval_s = 0.1
+                    pos_data = []
+                    for i in range(len(pos_lon)):
+                        cur_t = pos_t[i]
+                        if(cur_t - last_t) / 1e6 > minimum_interval_s:
+                            pos_data.append([pos_lat[i], pos_lon[i], current_flight_mode, \
```

Comment:
We can simplify this by keeping 2 lists: `pos_data` and `flight_modes` (that contains the color and indexes of pos_data where the new flight mode starts) and then pass them as separate jinja variables.
`pos_data.append([pos_lat[i], pos_lon[i])`

And then in the loop below:
```
                            while current_flight_mode_idx < len(flight_mode_changes) - 1 and \
                                    flight_mode_changes[current_flight_mode_idx + 1][0] <= cur_t:
                                current_flight_mode_idx += 1
                                current_flight_mode = flight_mode_changes[current_flight_mode_idx][1]
                                flight_modes.append([rgb_colors(current_flight_mode), i])
```

Then the java-script code will simplify as well to something like:
```
for(var j=0; j<flight_modes.length -1 ; j++)
{
    var waypoint_polyline = pos_polyline.slice(flight_modes[j][1], flight_modes[j+1][1]);
    var cur_flight_color = flight_modes[j][0];
    var polyline = L.polyline(waypoint_polyline, {color: cur_flight_color}).addTo(mymap);
    Array.prototype.push.apply(waypoint_bounds, waypoint_polyline);
}
```
Is this clear?